lib/ddtrace/contrib/rack/rum_injection.rb

```diff
@@ -0,0 +1,276 @@
+        CONTENT_LENGTH_HEADER = 'Content-Length'.freeze
+        CONTENT_TYPE_HEADER = 'Content-Type'.freeze
+        EXPIRES_HEADER = 'Expires'.freeze
+        SURROGATE_CACHE_CONTROL_HEADER = 'Surrogate-Control'.freeze
+        TRANSFER_ENCODING_HEADER = 'Transfer-Encoding'.freeze
+        ACTION_CONTROLLER_INSTANCE = 'action_controller.instance'.freeze
+        TRANSFER_ENCODING_CHUNKED = 'chunked'.freeze
+
+        def initialize(app)
+          @app = app
+        end
+
+        def call(env)
+          result = @app.call(env)
+
+          begin
+            return result unless configuration[:rum_injection_enabled] == true
+
+            status, headers, body = result
+
+            # need significant safety here to ensure result is parsable + injectable
+            # shouldn't be gzipped/compressed yet since we've injected our middleware in the stack after rack deflater
+            # or any other compression middleware for that matter
+            # also ensure its non-cacheable, is html, is not streaming, and is not an attachment
+            return result unless headers && should_inject?(headers, env)
+
+            trace_id = current_trace_id
+
+            # do not inject if no trace or trace is not sampled
+            return result unless trace_id
+
+            # update content length and return new response if we don't fail on rum injection
+            if body.respond_to?(:each)
+              # we need to insert the trace_id and expiry meta tags
+              unix_time = DateTime.now.strftime('%Q').to_i
+
+              html_comment = html_comment_template(trace_id, unix_time)
+
+              rum_body = RumBody.new(body, html_comment)
+
+              update_content_length(headers, html_comment)
+              # ensure idempotency on injection in case middleware is inserted or called twice
+              env[RUM_INJECTION_FLAG] = true
+
+              updated_response = ::Rack::Response.new(rum_body, status, headers)
+              Datadog.logger.debug { "Rum injection successful: #{html_comment}" }
+              return updated_response.finish
+            else
+              Datadog.logger.debug('Rum injection unsuccessful')
+              return result
+            end
+
+            # catchall if an earlier conditional is not met
+            result
+          rescue StandardError => e
+            Datadog.logger.warn("error checking rum injectability #{e.class}: #{e.message} #{e.backtrace.join("\n")}")
+            # we should ensure we don't interfere if original app response if our injection code has an exception
+            result
+          end
+        end
+
+        def self.inject_rum_data(supplied_env = nil)
+          tracer = Datadog.configuration[:rack][:tracer]
+          span = tracer.active_span
+
+          # only return trace id if sampled
+          trace_id = span && span.sampled ? span.trace_id : nil
+          unix_time = DateTime.now.strftime('%Q').to_i
```

Comment:
We can skip this date generation if `trace_id` is `nil`, let's move it inside the if block, or return earlier if `!tracer_id`.

Reply:
👍  updated